acceptancetests/assess_upgrade_series.py

```diff
@@ -0,0 +1,135 @@
+    args = (machine, series)
+    if agree:
+        args += ('--agree',)
+    client.juju('upgrade-series prepare', args)
+
+
+def upgrade_series_complete(client, machine):
+    args = (machine)
+    client.juju('upgrade-series complete', args)
+
+
+def do_release_upgrade(client, machine):
+    try:
+        output = client.get_juju_output(
+            'ssh', machine, 'sudo do-release-upgrade -f '
+            'DistUpgradeViewNonInteractive', timeout=3600)
+    except subprocess.CalledProcessError as e:
+        raise AssertionError(
+            "do-release-upgrade failed on {}: {}".format(machine, e))
+
+    log.info("do-release-upgrade response: ".format(output))
+
+
+def reboot_machine(client, machine):
+    try:
+        log.info("Restarting: {}".format(machine))
+        cmd = build_ssh_cmd(client, machine, 'sudo shutdown -r now && exit')
+        output = subprocess.check_output(cmd, stderr=subprocess.STDOUT)
+        log.info("Restarting machine output: {}\n".format(output))
+    except subprocess.CalledProcessError as e:
+        logging.info(
+            "Error running shutdown:\nstdout: %s\nstderr: %s",
+            e.output, getattr(e, 'stderr', None))
+
+    log.info("wait_for_started()")
+    client.wait_for_started()
+
+
+def build_ssh_cmd(client, machine, command):
+    ssh_opts = [
+        "-o", "User ubuntu",
+        "-o", "UserKnownHostsFile /dev/null",
+        "-o", "StrictHostKeyChecking no",
+        "-o", "PasswordAuthentication no",
+    ]
+
+    status = client.get_status()
+    machine_status = status.get_machine(machine)
+    cmd = ["ssh"] + ssh_opts + [machine_status['public-address']] + [command]
+    return cmd
+
+
+def assert_correct_series(client, machine, expected):
+    """Verify that juju knows the correct series for the machine"""
+    status = client.get_status()
+    status = status.get_machine(machine)
+    machine_series = status['series']
+    if machine_series is not expected:
+        raise JujuAssertionError(
+            "Machine {} series doesn't match the expected series: {}"
+            .format(machine, expected))
+
+
+def setup(client, start_series):
+    """ Deploy charms, there are several under ./repository """
+    charm_source = local_charm_path(
+        charm=charm_app, juju_ver=client.version)
+    _, deploy_complete = client.deploy(charm_source, series=start_series)
+    log.info("Deployed {} with {}".format(charm_app, start_series))
+    # Wait for the deployment to finish.
+    client.wait_for(deploy_complete)
+
+
+def parse_args(argv):
+    parser = argparse.ArgumentParser(description="Test juju update series.")
+    add_basic_testing_arguments(parser)
+    parser.add_argument('--from-series', default='xenial', dest='from_series',
+                        help='Series to start machine and units with')
+    parser.add_argument('--to-series', default='bionic', dest='to_series',
+                        help='Series to upgrade machine and units to')
```

Comment:
Note how the default test updates from `xenial` to `bionic`. If  we want to flex the init system conversion functionality the test would have to be run with something like `--to-series=trusty --from-series=xenial`. Should `trusty->bionic` instead be the default.

Reply:
--to-series=_xenial_ --from-series=_trusty_